torch_agents/cogment_verse_torch_agents/muzero/adapter.py

```diff
@@ -0,0 +1,801 @@
+    priority_update_queue = mp.Queue()
+    reanalyze_update_queue = mp.Queue()
+    reanalyze_queue = mp.Queue(num_reanalyze_workers + 1)
+
+    batch_queue = mp.Queue(max_prefetch_batch)
+    reanalyze_queue = mp.Queue()
+    agent_update_queue = mp.Queue()
+    # limit to small size so that training and sample generation don't get out of sync
+    info_queue = mp.Queue(max_prefetch_batch)
+
+    agent_queue = mp.Queue()
+    reanalyze_agent_queue = mp.Queue()
+
+    reward_distribution = copy.deepcopy(agent.muzero.reward_distribution).cpu()
+    value_distribution = copy.deepcopy(agent.muzero.value_distribution).cpu()
+
+    train_worker = TrainWorker(agent, batch_queue, agent_update_queue, info_queue, config)
+    replay_buffer = ReplayBufferWorker(
+        sample_queue,
+        priority_update_queue,
+        batch_queue,
+        reanalyze_queue,
+        reanalyze_update_queue,
+        config.training,
+        reward_distribution,
+        value_distribution,
+    )
+    reanalyze_agent_queue = [mp.Queue() for _ in range(num_reanalyze_workers)]
+    reanalyze_workers = [
+        ReanalyzeWorker(
+            reanalyze_queue,
+            reanalyze_update_queue,
+            reanalyze_agent_queue[i],
+            config.training.reanalyze_device,
+            reward_distribution,
+            value_distribution,
+        )
+        for i in range(num_reanalyze_workers)
+    ]
+
+    trials_completed = 0
+    running_stats = RunningStats()
+
+    xp_tracker.log_params(
+        config.environment,
+        config.actor,
+        config.training,
+    )
+
+    trial_configs = make_trial_configs(run_session, config, model_id, -1)
+
+    total_samples = 0
+    training_step = 0
+    run_total_reward = 0
+
+    workers = [train_worker, replay_buffer] + reanalyze_workers
+
+    try:
+        for worker in workers:
+            worker.start()
+
+        start_time = time.time()
+
+        sample_generator = run_session.start_trials_and_wait_for_termination(
+            trial_configs,
+            max_parallel_trials=config.training.max_parallel_trials,
+        )
+
+        start_time = time.time()
+        samples = 0
+
+        async for _step, timestamp, trial_id, _tick, (sample, total_reward) in sample_generator:
+            samples += 1
+            replay_buffer.add_sample(trial_id, sample)
+            total_samples += 1
+
+            for worker in workers:
+                assert worker.is_alive()
+
+            if sample.done:
+                trials_completed += 1
+                xp_tracker.log_metrics(
+                    timestamp, total_samples, trial_total_reward=total_reward, trials_completed=trials_completed
+                )
+                run_total_reward += total_reward
+
+            if replay_buffer.size() <= config.training.min_replay_buffer_size:
+                continue
+
+            # get latest online model
+            try:
+                agent = agent_update_queue.get_nowait()
```

Comment:
Do you actually need the agent_update_queue ? can't you just provide a list of outputs to the train worker and avoid this polling ?

Reply:
Metrics and model updates now merged into a single `results_queue`